### ISSUE TITLE
NeedPrivileges Exception for CalDAV url

### ISSUE THREAD

**Issue by johnwoltman** [owncloud/calendar#782]
Upgraded to 8.0.3 and Calendar 0.6.4, and encountered this issue:

Logging in as an LDAP user or as the owncloud administrator account, I cannot view/sync either CardDAV or CalDAV data.  I can use the WebDAV support just fine, but not the Card/CalDAV. 

### Expected behaviour
Should be able to use CardDAV and CalDAV links with Outlook or view it with Firefox.

### Actual behaviour
Receive Sabre\DAVACL\Exception\NeedPrivileges exception

### Server configuration
**Operating system**:Centos 7.1 x86_64

**Web server:** Apache 2.4.6

**Database:**  postgresql 9.3.6

**PHP version:** 5.6.8

**ownCloud version:** 8.0.3

**Updated from an older ownCloud or fresh install:** Update from 7

**List of activated apps:**
 - activity 1.2.0    
 - calendar 0.6.4    
 - contacts 0.3.0.18
 - files             
 - files_sharing
 - files_texteditor  
 - files_trashbin    
 - files_versions    
 - files_videoviewer 
 - gallery           
 - news 5.2.8       
 - user_ldap 0.5.0 

**The content of config/config.php:**

```
<?php
$CONFIG = array (
  'instanceid' => '51d4672e6805c',
  'passwordsalt' => '[redacted]',
  'datadirectory' => '/var/cloud',
  'dbtype' => 'pgsql',
  'version' => '8.0.3.4',
  'dbname' => 'owncloud',
  'dbhost' => '127.0.0.1',
  'dbtableprefix' => 'oc_',
  'dbuser' => '[redacted]',
  'dbpassword' => '[redacted]',
  'installed' => true,
  'ldapIgnoreNamingRules' => false,
  'loglevel' => 2,
  'log_type' => 'syslog',
  'logfile' => '',
  'maintenance' => false,
  'theme' => '',
  'forcessl' => true,
  'appcodechecker' => false,
  'overwritewebroot' => '/cloud',
  'overwrite.cli.url' => '/cloud',
  'trusted_domains' =>
  array (
    0 => '[redacted]',
  ),
  'secret' => '[redacted]',
);
```

**Are you using external storage, if yes which one:** No

**Are you using encryption:** Only SSL/TLS, not owncloud file encryption

**Are you using an external user-backend, if yes which one:** LDAP/

#### LDAP configuration (delete this part if not used)

```
   appid   |             configkey             |                 configvalue
-----------+-----------------------------------+---------------------------------------------
 user_ldap | ldap_agent_password               | [redacted]
 user_ldap | ldap_turn_off_cert_check          |
 user_ldap | bgjUpdateGroupsLastRun            | 1386377150
 user_ldap | ldap_quota_attr                   |
 user_ldap | ldap_backup_host                  |
 user_ldap | ldap_attributes_for_group_search  |
 user_ldap | enabled                           | yes
 user_ldap | ldap_group_filter_mode            | 0
 user_ldap | ldap_cache_ttl                    | 600
 user_ldap | ldap_groupfilter_objectclass      | inetOrgPerson
 user_ldap | ldap_user_filter_mode             | 0
 user_ldap | ldap_group_filter                 | (&(|(objectclass=inetOrgPerson)))
 user_ldap | ldap_backup_port                  | 389
 user_ldap | ldap_configuration_active         | 1
 user_ldap | ldap_login_filter                 | (&(|(objectclass=inetOrgPerson))(uid=%uid))
 user_ldap | ldap_base_groups                  | ou=users,dc=[redacted],dc=com
 user_ldap | ldap_attributes_for_user_search   |
 user_ldap | ldap_tls                          |
 user_ldap | ldap_nocase                       |
 user_ldap | ldap_override_main_server         |
 user_ldap | ldap_quota_def                    |
 user_ldap | ldap_display_name                 | cn
 user_ldap | ldap_userfilter_objectclass       | inetOrgPerson
 user_ldap | home_folder_naming_rule           |
 user_ldap | ldap_expert_username_attr         | cn
 user_ldap | ldap_dn                           | cn=[redacted],dc=[redacted],dc=com
 user_ldap | ldap_port                         | 389
 user_ldap | ldap_expert_uuid_attr             |
 user_ldap | ldap_uuid_attribute               | entryuuid
 user_ldap | types                             | authentication
 user_ldap | ldap_host                         | ldap://[redacted]
 user_ldap | ldap_base                         | dc=[redacted],dc=com
 user_ldap | ldap_email_attr                   | mail
 user_ldap | ldap_nested_groups                | 0
 user_ldap | ldap_paging_size                  | 500
 user_ldap | ldap_base_users                   | ou=users,dc=[redacted],dc=com
 user_ldap | ldap_userfilter_groups            |
 user_ldap | ldap_userlist_filter              | (|(objectclass=inetOrgPerson))
 user_ldap | ldap_groupfilter_groups           |
 user_ldap | ldap_group_display_name           | cn
 user_ldap | ldap_group_member_assoc_attribute | uniqueMember
 user_ldap | ldap_login_filter_mode            | 0
 user_ldap | ldap_loginfilter_email            | 0
 user_ldap | ldap_loginfilter_username         | 1
 user_ldap | ldap_loginfilter_attributes       |
 user_ldap | has_memberof_filter_support       |
 user_ldap | ldap_expert_uuid_user_attr        |
 user_ldap | ldap_expert_uuid_group_attr       |
 user_ldap | last_jpegPhoto_lookup             | 0
 user_ldap | ldap_uuid_user_attribute          | entryuuid
 user_ldap | ldap_uuid_group_attribute         | entryuuid
 user_ldap | ocsid                             | 166061
 user_ldap | ldap_experienced_admin            | 0
 user_ldap | installed_version                 | 0.5.0
 user_ldap | cleanUpJobOffset                  | 0
```

### Client configuration
**Browser:** Firefox 36

**Operating system:** Windows 7 64-bit

#### Web server & ownCloud error log
```
Apr 15 01:06:20 [redacted] ownCloud[1768]: {news} Configuration value "simplePieCacheDuration" does not exist. Ignored value.
Apr 15 01:06:20 [redacted] ownCloud[1824]: {news} Configuration value "simplePieCacheDuration" does not exist. Ignored value.
Apr 15 01:13:25 [redacted] ownCloud[6242]: {PHP} Undefined index: REQUEST_URI at /var/www/html/owncloud/apps/calendar/appinfo/app.php#41
Apr 15 01:13:29 [redacted] ownCloud[6245]: {PHP} Undefined index: REQUEST_URI at /var/www/html/owncloud/apps/calendar/appinfo/app.php#41
Apr 15 01:13:34 [redacted] ownCloud[6250]: {PHP} Undefined index: REQUEST_URI at /var/www/html/owncloud/apps/calendar/appinfo/app.php#41
Apr 15 01:20:02 [redacted] ownCloud[6360]: {PHP} Undefined index: REQUEST_URI at /var/www/html/owncloud/apps/calendar/appinfo/app.php#41
Apr 15 01:20:02 [redacted] ownCloud[6360]: {news} Configuration value "simplePieCacheDuration" does not exist. Ignored value.
Apr 15 01:35:41 [redacted] ownCloud[5713]: PHP Fatal error:  Class OC_Connector_Sabre_Auth contains 1 abstract method and must therefore be declared abstract or implement the remaApr 15 01:35:41 [redacted] httpd[5713]: {PHP} Class OC_Connector_Sabre_Auth contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (Apr 15 01:36:17 [redacted] httpd[6727]: [Wed Apr 15 01:36:17.474944 2015] [so:warn] [pid 6727] AH01574: module wsgi_module is already loaded, skipping
Apr 15 01:36:17 [redacted] httpd[6727]: AH00548: NameVirtualHost has no effect and will be removed in the next release /etc/httpd/conf.d/com.conf:1
Apr 15 01:40:02 [redacted] ownCloud[7071]: {PHP} Undefined index: REQUEST_URI at /var/www/html/owncloud/apps/calendar/appinfo/app.php#41
Apr 15 01:40:02 [redacted] ownCloud[7071]: {news} Configuration value "simplePieCacheDuration" does not exist. Ignored value.
Apr 15 01:40:53 [redacted] ownCloud[6786]: {carddav} Exception: {"Message":"User did not have the required privileges ({DAV:}read) for path \"addressbooks\/john\/contacts\"","CodeApr 15 01:42:01 [redacted] httpd[7161]: [Wed Apr 15 01:42:01.133809 2015] [so:warn] [pid 7161] AH01574: module wsgi_module is already loaded, skipping

Apr 15 01:42:06 [redacted] ownCloud[7165]: {carddav} Exception: {"Message":"User did not have the required privileges ({DAV:}read) for path \"addressbooks\/john\/contacts\"","CodeApr 15 01:46:34 [redacted] ownCloud[7302]: {PHP} Undefined index: REQUEST_URI at /var/www/html/owncloud/apps/calendar/appinfo/app.php#41
```

#### Browser log
```
None
```

**Comment by georgehrke** [comment]
cc @blizzz any idea?

**Comment by RealRancor** [comment]
Its still not clear if this is an issue in either the calendar/contacts apps or core. Some refs: https://github.com/owncloud/calendar/issues/776, https://github.com/owncloud/core/issues/14048, https://github.com/owncloud/contacts/issues/814

**Comment by RealRancor** [comment]
Btw. in core a fix was applied with https://github.com/owncloud/core/pull/14275 to solve a similar message. Maybe there is something similar needed for the calendar and contacts apps?

**Comment by johnwoltman** [comment]
I can get on IRC if anyone wants me to help them track down this bug.  I'll be available Tues-Thurs this week week during late afternoons (Pacific time zone). I'm eager to help to get this issue squashed.

**Comment by monnimeter** [comment]
Same here - I cannot sync my calendar with CalDAV.
But, I do not use the LDAP backend, so it seems there could be something else wrong.

While testing around I've found this behaviour:

If I trigger a request using this URL (from Firefox)
https://192.168.178.2/owncloud/remote.php/caldav/calendars/xxxxx@xxxxx.at/defaultcalendar
then I get this entry in the log file:

    Exception: {"Message":"HTTP\/1.1 403 User did not have the required privileges ({DAV:}read) for path \"calendars\/xxxxx@xxxxx.at\/defaultcalendar\"","Code":0,"Trace":"#0 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAVACL\/Plugin.php(769): Sabre\\DAVACL\\Plugin->checkPrivileges('calendars\/osman...', '{DAV:}read')\n#1 [internal function]: Sabre\\DAVACL\\Plugin->beforeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#2 \/srv\/www\/owncloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\n#3 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(456): Sabre\\Event\\EventEmitter->emit('beforeMethod', Array)\n#4 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#5 \/srv\/www\/owncloud\/apps_usr\/calendar\/appinfo\/remote.php(57): Sabre\\DAV\\Server->exec()\n#6 \/srv\/www\/owncloud\/remote.php(84): require_once('\/srv\/www\/ownclo...')\n#7 {main}","File":"\/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAVACL\/Plugin.php","Line":223} 

But, if I change the URL to (just substitute the IP adress with the host name) to
https://hpstorage.locl/owncloud/remote.php/caldav/calendars/xxxxx@xxxxx.at/defaultcalendar
then the error changes to:

    Exception: {"Message":"HTTP\/1.1 501 There was no handler found for this \"GET\" method","Code":0,"Trace":"#0 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#1 \/srv\/www\/owncloud\/apps_usr\/calendar\/appinfo\/remote.php(57): Sabre\\DAV\\Server->exec()\n#2 \/srv\/www\/owncloud\/remote.php(84): require_once('\/srv\/www\/ownclo...')\n#3 {main}","File":"\/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","Line":472} 

Tested on: Owncloud Version 8.1 beta 1
Calendar 0.6.4

THX and regards,
Monni

**Comment by RealRancor** [comment]
> But, I do not use the LDAP backend, so it seems there could be something else wrong.

Are you using any other user backends as you have a username of xxxxx@xxxxx.at ?

**Comment by monnimeter** [comment]
No, just DB.

This is what I've installed so far:

    root@hpstorage:/srv/www/owncloud# sudo -u www-data php occ app:list
Enabled:
 - activity: 1.2.3
 - calendar: 0.6.4
 - contacts: 0.3.0.18
 - documents: 0.9.0
 - external: 1.2
 - files: 1.1.9
 - files_antivirus: 0.6.1
 - files_embeddedvideo: 0.0.7
 - files_external: 0.2.3
 - files_mv: 0.8.1
 - files_pdfviewer: 0.7
 - files_sharing: 0.6.1
 - files_texteditor: 0.4
 - files_videoviewer: 0.1.3
 - firstrunwizard: 1.1
 - galleryplus: 8.1.11
 - mail: 0.1.3
 - mozilla_sync: 1.4
 - music: 0.3.5
 - notes: 1.0.0
 - search_lucene: 0.6.0
 - tasks: 0.6.0
 - templateeditor: 0.1
 - updater: 0.4
 - user_external: 0.4
Disabled:
 - encryption
 - files_locking
 - files_trashbin
 - files_versions
 - gallery
 - provisioning_api
 - user_ldap
 - user_webdavauth

**Comment by monnimeter** [comment]
Sorry for the delay. This is what happens:

URL: https://192.168.178.2/owncloud/remote.php/caldav/calendars/xxxxx@xxxxx.at
Error:

    Exception: {"Message":"HTTP\/1.1 403 User did not have the required privileges ({DAV:}read) for path \"calendars\/xxxxx@xxxxx.at\"","Code":0,"Trace":"#0 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAVACL\/Plugin.php(769): Sabre\\DAVACL\\Plugin->checkPrivileges('calendars\/xxxxx...', '{DAV:}read')\n#1 [internal function]: Sabre\\DAVACL\\Plugin->beforeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#2 \/srv\/www\/owncloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\n#3 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(456): Sabre\\Event\\EventEmitter->emit('beforeMethod', Array)\n#4 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#5 \/srv\/www\/owncloud\/apps_usr\/calendar\/appinfo\/remote.php(57): Sabre\\DAV\\Server->exec()\n#6 \/srv\/www\/owncloud\/remote.php(84): require_once('\/srv\/www\/ownclo...')\n#7 {main}","File":"\/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAVACL\/Plugin.php","Line":223} 

URL: https://hpstorage.locl/owncloud/remote.php/caldav/calendars/xxxxx@xxxxx.at
Error:

    Exception: {"Message":"HTTP\/1.1 501 There was no handler found for this \"GET\" method","Code":0,"Trace":"#0 \/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))\n#1 \/srv\/www\/owncloud\/apps_usr\/calendar\/appinfo\/remote.php(57): Sabre\\DAV\\Server->exec()\n#2 \/srv\/www\/owncloud\/remote.php(84): require_once('\/srv\/www\/ownclo...')\n#3 {main}","File":"\/srv\/www\/owncloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","Line":472}

**Comment by childnode** [comment]
that you see this error, it's no error for me but a feature ;)
because calDAV/cardDAV/webDAV interface 
* has/defines and uses mostly it's own HTTP methods, not GET and POST as you usally do implicit on Websites
* s.th. GET readable is not at the endpoint adress addressboox/username/calendarname or principal/username itself but on its subpath /calendar-proxy-read and /calendar-proxy-read
 * have a look at what PROPFIND returns

What I don't understand: why is the GET readable content empty?

even cardDav is not defined to respond to GET requests but PROPFIND too
(see https://tools.ietf.org/html/rfc6352 || http://sabre.io/dav/building-a-carddav-client/ )

What is wrong to me there:
```
<d:response>
  <d:href>/remote.php/carddav/addressbooks/marcel/kontakte</d:href>
  <d:propstat>
    <d:prop>
      <d:getlastmodified/>
      <d:getcontentlength/>
      <d:resourcetype/>
      <d:quota-used-bytes/>
      <d:quota-available-bytes/>
      <d:getetag/>
      <d:getcontenttype/>
    </d:prop>
    <d:status>HTTP/1.1 403 Forbidden</d:status>
  </d:propstat>
</d:response>
```

followed by man vcf entries like 
`/remote.php/carddav/addressbooks/marcel/kontakte/137755f1-8973-4a11-8d34-8d6262dd614f%2540owncloud.childno.de.vcf`
which should be accessible by GET but it isn't.

While complete list / VCF of alle elements is correctly returned by REPORT.

Back to CalDAV
http://sabre.io/dav/building-a-caldav-client/
analogusly here, PROPFIND returns a much more simple list with to 200 OK status links to /calendar-proxy-read and /calendar-proxy-read#


For me:
* Calendar Sync is working fine to me with
 * DavDroid 0.77
 * Mac OSX 10.9
* Contact Sync is working fine to me with
 * DavDroid

BUT Contact sync fails on Mac.. the manual https://doc.owncloud.org/server/4.5/user_manual/sync_osx.html using /principal paths (which are somehow determined automatically by the Apple Account setup) 
Perhaps I have to read more technical documentation to find the correct "issues".

**To sum up:** don't bother about this message. Check what a DAV Client is telling about.

**Comment by t3dev** [comment]
same issue here after upgrade of owncloud 8.1.1-1 debian package from download.opensuse.org

**Comment by jfburdet** [comment]
This bug is eavily impacting my users. I can provide debug info if you give me guidelines.
